src/smt/theory_str_noodler/theory_str_noodler_final_check.cpp

```diff
@@ -535,6 +535,135 @@ namespace smt::noodler {
         return l_undef;
     }
 
+    bool theory_str_noodler::is_mult_membership_suitable() {
+        if (!this->m_conversion_todo.empty() && !this->m_not_contains_todo_rel.empty()) {
+            return false;
+        }
+
+        // TODO handle (string_literal in RE) and also length vars, instead of giving up
+        for (const auto& membership: m_membership_todo_rel) {
+            if(m_util_s.str.is_string(to_app(std::get<0>(membership))) || len_vars.contains(std::get<0>(membership))) {
+                return false;
+            }
+        }
+
+        // we now check if we only have "trivial" (dis)equations, i.e. they can only be of form
+        //    x == str_literal or x != str_literal
+        // i.e., one variable on the left, a string literal on the right
+
+        for (const auto& diseq : m_word_diseq_todo_rel) {
+            if (!(util::is_str_variable(diseq.first, m_util_s) && m_util_s.str.is_string(diseq.second))
+                || len_vars.contains(diseq.first)) // TODO handle length vars in run_mult_membership_heur() by using the lengths from the final intersection automaton
+            {
+                return false;
+            }
+        }
+
+        for (const auto& eq : m_word_eq_todo_rel) {
+            if (!(util::is_str_variable(eq.first, m_util_s) && m_util_s.str.is_string(eq.second))
+                || len_vars.contains(eq.first)) // TODO handle length vars in run_mult_membership_heur() by using the lengths from the final intersection automaton
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    lbool theory_str_noodler::run_mult_membership_heur() {
+        STRACE("str", tout << "trying multiple regex membership heuristic" << std::endl;);
+
+        // gets the sum of loops of a regex (loop inside a loop is multiplied)
+        std::function<unsigned(app*)> get_loop_sum;
+        get_loop_sum = [this, &get_loop_sum](app* reg) -> unsigned {
```

Comment:
make it as a function in `smt::noodler::regex` (`regex.h`). It seems to be quite reusable.

Reply:
Done